provider/vsphere/internal/vsphereclient/createvm.go

```diff
@@ -133,78 +136,100 @@ func (c *Client) CreateVirtualMachine(
+		return nil, errors.Trace(err)
+	}
+	datastore := object.NewDatastore(c.client.Client, datastoreMo.Reference())
+	datastore.DatacenterPath = datacenter.InventoryPath
+	datastore.SetInventoryPath(path.Join(folders.DatastoreFolder.InventoryPath, datastoreMo.Name))
+
+	// Ensure the VMDK is present in the datastore, uploading it if it
+	// doesn't already exist.
+	resourcePool := object.NewResourcePool(c.client.Client, *args.ComputeResource.ResourcePool)
+	taskWaiter := &taskWaiter{args.Clock, args.UpdateProgress, args.UpdateProgressInterval}
+	vmdkDatastorePath, releaseVMDK, err := c.ensureVMDK(ctx, args, datastore, datacenter, taskWaiter)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	defer releaseVMDK()
+
+	// Import the VApp, creating a temporary VM. This is necessary to
+	// import the VMDK, which exists in the datastore as a not-a-disk
+	// file type.
+	args.UpdateProgress("creating import spec")
+	importSpec, err := c.createImportSpec(ctx, args, datastore, vmdkDatastorePath)
+	if err != nil {
+		return nil, errors.Annotate(err, "creating import spec")
+	}
+	importSpec.ConfigSpec.Name += ".tmp"
+
 	args.UpdateProgress(fmt.Sprintf("creating VM %q", args.Name))
-	c.logger.Debugf("creating VM in folder %s", vmFolder)
-	rp := object.NewResourcePool(c.client.Client, *args.ComputeResource.ResourcePool)
-	lease, err := rp.ImportVApp(ctx, spec.ImportSpec, vmFolder, nil)
+	c.logger.Debugf("creating temporary VM in folder %s", vmFolder)
+	c.logger.Tracef("import spec: %s", pretty.Sprint(importSpec))
+	lease, err := resourcePool.ImportVApp(ctx, importSpec, vmFolder, nil)
 	if err != nil {
 		return nil, errors.Annotatef(err, "failed to import vapp")
 	}
-
-	// Upload the VMDK.
-	info, err := lease.Wait(ctx)
+	info, err := lease.Wait(ctx, nil)
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
-	type uploadItem struct {
-		item types.OvfFileItem
-		url  *url.URL
+	if err := lease.Complete(ctx); err != nil {
+		return nil, errors.Trace(err)
 	}
-	var uploadItems []uploadItem
-	for _, device := range info.DeviceUrl {
-		for _, item := range spec.FileItem {
-			if device.ImportKey != item.DeviceId {
-				continue
-			}
-			u, err := c.client.Client.ParseURL(device.Url)
-			if err != nil {
-				return nil, errors.Trace(err)
-			}
-			uploadItems = append(uploadItems, uploadItem{
-				item: item,
-				url:  u,
-			})
+	tempVM := object.NewVirtualMachine(c.client.Client, info.Entity)
+	defer func() {
+		if err := c.destroyVM(ctx, tempVM, taskWaiter); err != nil {
+			c.logger.Warningf("failed to delete temporary VM: %s", err)
+		}
+	}()
+
+	// Clone the temporary VM to import the VMDK, as mentioned above.
+	// After cloning the temporary VM, we must detach the original
+	// VMDK from the temporary VM to avoid deleting it when destroying
+	// the VM.
+	c.logger.Debugf("cloning VM")
+	vm, err := c.cloneVM(ctx, tempVM, args.Name, vmFolder, taskWaiter)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	defer func() {
+		if err == nil {
```

Comment:
I can't quite follow what err this is referencing.
Shouldn't we use a named return error for the entire method and this err would then shadow that return error value?

Reply:
done